### PR TITLE
Created setVoltage and implemented for tankDrive and arcadeDrive

### PR DESCRIPTION
**Pull request by Yoel-Perman-Brilliant**

### CODE REVIEW

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+public class DriveTankWithVoltage extends CommandBase {
+    protected final Supplier<Double> leftVoltage;
+    protected final Supplier<Double> rightVoltage;
+    protected final TankDrivetrain drivetrain;
+    protected final Supplier<Boolean> isFinished;
+
+    /**
+     * This constructs a new {@link DriveTank} command that moves the given
+     * {@link TankDrivetrain} according to speed values from Double {@link Supplier}s for left and right sides.<br>
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java:22]: voltage values

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     * Positive values move forwards.
+     *
+     * @param drivetrain   the drivetrain this command requires and moves.
+     * @param leftVoltage  the double {@link Supplier} supplying the voltage to move on the right side with.
+     * @param rightVoltage the double {@link Supplier} supplying the voltage to move on the left side with.
+     * @param isFinished
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java:28]: where description

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    public DriveTankWithVoltage(TankDrivetrain drivetrain, Supplier<Double> leftVoltage, Supplier<Double> rightVoltage, Supplier<Boolean> isFinished) {
+        this.drivetrain =drivetrain;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java:31]: A-II\a6

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        this.leftVoltage=leftVoltage;
+        this.rightVoltage=rightVoltage;
+        this.isFinished = isFinished;
+    }
+    @Override
+    public void execute() {
+        drivetrain.tankDriveVoltages(leftVoltage.get(), rightVoltage.get());
+    }
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java:39]: A-II\a10

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    public boolean isFinished() {
+        return this.isFinished.get();
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java:41]: why this?????

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltage.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package com.spikes2212.command.drivetrains.commands;
+
+import java.util.function.Supplier;
+
+import com.spikes2212.command.drivetrains.TankDrivetrain;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+
+/**
+ * This command moves a {@link TankDrivetrain} by linear and rotational speeds, using
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltage.java:10]: voltages

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltage.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
+package com.spikes2212.command.drivetrains.commands;
+
+import java.util.function.Supplier;
+
+import com.spikes2212.command.drivetrains.TankDrivetrain;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltage.java:7]: ?

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.util.function.Supplier;
+
+/**
+ * A command that moves a {@link TankDrivetrain} using voltage supplied to the left and the right sides independently.
+ *
+ * @author Yuval Levy
+ * @see TankDrivetrain
+ */
+public class DriveTankWithVoltage extends CommandBase {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltage.java:14]: A-II\b3

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+     * {@link TankDrivetrain} according to speed values from Double {@link Supplier}s for left and right sides.<br>
+     * Positive values move forwards.
+     *
+     * @param drivetrain   the drivetrain this command requires and moves.
+     * @param leftVoltage  the double {@link Supplier} supplying the voltage to move on the right side with.
+     * @param rightVoltage the double {@link Supplier} supplying the voltage to move on the left side with.
+     * @param isFinished
+     */
+    public DriveTankWithVoltages(TankDrivetrain drivetrain, Supplier<Double> leftVoltage, Supplier<Double> rightVoltage, Supplier<Boolean> isFinished) {
+        this.drivetrain =drivetrain;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:31]: CTRL ALT L

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        this.leftVoltage=leftVoltage;
+        this.rightVoltage=rightVoltage;
+        this.isFinished = isFinished;
+    }
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:35]: blank line

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    @Override
+    public void execute() {
+        drivetrain.tankDriveVoltages(leftVoltage.get(), rightVoltage.get());
+    }
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:39]: blank line

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltages.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    protected final TankDrivetrain tankDrivetrain;
+    protected final Supplier<Double> moveValueSupplier;
+    protected final Supplier<Double> rotateValueSupplier;
+    protected final Supplier<Boolean> isFinished;
+
+    /**
+     * This constructs a new {@link DriveArcadeWithVoltages} command that moves the given
+     * {@link TankDrivetrain} according to Voltage values for Linear and rotational movements.
+     *
+     * @param drivetrain          the tank drivetrain this command operates on.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltages.java:27]: remove period

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltages.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+     * @param moveValueSupplier   the double {@link Supplier} supplying the linear Voltage. Positive values go forwards.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltages.java:28]: remove final period

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltages.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+     * @param rotateValueSupplier the double {@link Supplier} supplying the rotational Voltage. Positive values go left.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithVoltages.java:29]: remove final period

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.spikes2212.command.drivetrains.TankDrivetrain;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+import java.util.function.Supplier;
+
+/**
+ * A command that moves a {@link TankDrivetrain} using voltage supplied to the left and the right sides independently.
+ *
+ * @author Yuval Levy
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:11]: NO

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
+package com.spikes2212.command.drivetrains.commands;
+
+import com.spikes2212.command.drivetrains.TankDrivetrain;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+import java.util.function.Supplier;
+
+/**
+ * A command that moves a {@link TankDrivetrain} using voltage supplied to the left and the right sides independently.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:9]: rephrase to be more in line with documentation for other commands

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @author Yuval Levy
+ * @see TankDrivetrain
+ */
+public class DriveTankWithVoltages extends CommandBase {
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:14]: add blank line after this one

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * This constructs a new {@link DriveTank} command that moves the given
+     * {@link TankDrivetrain} according to speed values from Double {@link Supplier}s for left and right sides.<br>
+     * Positive values move forwards.
+     *
+     * @param drivetrain   the drivetrain this command requires and moves.
+     * @param leftVoltage  the double {@link Supplier} supplying the voltage to move on the right side with.
+     * @param rightVoltage the double {@link Supplier} supplying the voltage to move on the left side with.
+     * @param isFinished
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:28]: ???

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    protected final Supplier<Double> rightVoltage;
+    protected final TankDrivetrain drivetrain;
+    protected final Supplier<Boolean> isFinished;
+
+    /**
+     * This constructs a new {@link DriveTank} command that moves the given
+     * {@link TankDrivetrain} according to speed values from Double {@link Supplier}s for left and right sides.<br>
+     * Positive values move forwards.
+     *
+     * @param drivetrain   the drivetrain this command requires and moves.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:25]: remove period

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+     * @param leftVoltage  the double {@link Supplier} supplying the voltage to move on the right side with.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:26]: remove period

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+     * @param rightVoltage the double {@link Supplier} supplying the voltage to move on the left side with.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:27]: remove period

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+    public boolean isFinished() {
+        return isFinished.get();
+    }
+
+    @Override
+    public void end(boolean interrupted) {
+        drivetrain.stop();
+    }
+
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveTankWithVoltages.java:50]: delete this blank line

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]

File: src/main/java/com/spikes2212/command/drivetrains/TankDrivetrain.java [hunk 23]
```diff
@@ -95,7 +119,7 @@ public void curvatureDrive(double speed, double curvature) {
     /**
      * Moves the left side of this drivetrain by a given speed.
      *
-     * @param speedLeft the speed to set to the left side. Positive values move this side forward.
+     * @param speedLeft the speed to set to the left side (range from -1 to 1). Positive values move this side forward
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/TankDrivetrain.java:122]: just -1 to 1, it's more consistent with the rest of the documentations

**Review by Spikes-2212 — APPROVED** [review]

**Pull request merged**